test/distributed/algorithms/test_join.py

```diff
@@ -0,0 +1,420 @@
+
+        inputs = self.construct_uneven_inputs(BASE_NUM_INPUTS, OFFSET)
+        allreduce_total = 0
+        with _Join(allreducers, run_post_hooks=True):
+            for _ in inputs:
+                for allreducer in allreducers:
+                    allreduce_total += allreducer()
+
+        expected_total = self.world_size * BASE_NUM_INPUTS
+        for num_joined in range(1, self.rank + 1):
+            expected_total += (self.world_size - num_joined) * OFFSET
+        # The expected total is now multiplied by a factor of `NUM_JOINABLES`
+        expected_total *= NUM_JOINABLES
+        self.assertEqual(allreduce_total, expected_total)
+        for allreducer in allreducers:
+            self.assertEqual(allreducer.post_hook_tensor.item(), AFTER_CONSTANT)
+
+    @require_n_gpus_for_nccl_backend(
+        WORLD_SIZE, BACKEND
+    )
+    def test_single_joinable_disable(self):
+        r"""Tests ``enable=False`` for a single :class:`_Joinable`."""
+        self.dist_init(self.rank, self.world_size)
+        BASE_NUM_INPUTS = 3
+
+        allreducer = AllReducer(self.device, self.process_group)
+        inputs = self.construct_even_inputs(BASE_NUM_INPUTS)
+        allreduce_total = 0
+        with _Join([allreducer], enable=False):
+            for _ in inputs:
+                allreduce_total += allreducer()
+
+        expected_total = self.world_size * BASE_NUM_INPUTS
+        self.assertEqual(allreduce_total, expected_total)
+
+    @require_n_gpus_for_nccl_backend(
+        WORLD_SIZE, BACKEND
+    )
+    def test_multiple_joinable_disable(self):
+        r"""
+        Tests ``enable=False`` for multiple :class:`_Joinable` s.
+
+        This generalizes ``test_single_joinable_disable`` to multiple
+        :class:`_Joinable` s.
+        """
+        self.dist_init(self.rank, self.world_size)
+        BASE_NUM_INPUTS = 3
+        NUM_JOINABLES = 3
+
+        allreducers = [
+            AllReducer(self.device, self.process_group)
+            for _ in range(NUM_JOINABLES)
+        ]
+        inputs = self.construct_even_inputs(BASE_NUM_INPUTS)
+        allreduce_total = 0
+        with _Join(allreducers, enable=False):
+            for _ in inputs:
+                for allreducer in allreducers:
+                    allreduce_total += allreducer()
+
+        expected_total = self.world_size * BASE_NUM_INPUTS * NUM_JOINABLES
+        self.assertEqual(allreduce_total, expected_total)
+
+    @require_n_gpus_for_nccl_backend(
+        WORLD_SIZE, BACKEND
+    )
+    def test_single_joinable_throw(self):
+        r"""
+        Tests ``throw_on_early_termination=True`` for a single
+        :class:`_Joinable`.
+        """
+        self.dist_init(self.rank, self.world_size)
+        BASE_NUM_INPUTS = 3
+        OFFSET = 1
+
+        allreducer = AllReducer(self.device, self.process_group)
+        inputs = self.construct_uneven_inputs(BASE_NUM_INPUTS, OFFSET)
+        allreduce_total = 0
+        with self.assertRaises(RuntimeError):
```

Comment:
Shall we check the error message is expected?

Reply:
Done.